ZOOKEEPER-3309: Add sync processor metrics

jhuan31 opened the pull request.

eolivelli commented on zookeeper-server/src/main/java/org/apache/zookeeper/server/SyncRequestProcessor.java:
```diff
@@ -175,9 +182,14 @@ private void flush(LinkedList<Request> toFlush)
         if (toFlush.isEmpty())
             return;
 
+        LOG.info("flush size {}", toFlush.size());
```
Do we need this at info level? Is it useful?
Only asking

jhuan31 replied:
removed

eolivelli commented on zookeeper-server/src/test/java/org/apache/zookeeper/server/quorum/SyncRequestProcessorMetricTest.java:
```diff
@@ -0,0 +1,82 @@
+        when(zks.getZKDatabase()).thenReturn(db);
+    }
+
+    @After
```
Can we drop this empty method?

jhuan31 replied:
removed

eolivelli commented on zookeeper-server/src/test/java/org/apache/zookeeper/server/quorum/SyncRequestProcessorMetricTest.java:
```diff
@@ -0,0 +1,82 @@
+
+        try {
+            Thread.sleep(500);
+        } catch (Exception e){
```
Why?
If you get an IE the test will fail, not a big deal

jhuan31 replied:
modified

lvfangmin commented on zookeeper-server/src/main/java/org/apache/zookeeper/server/SyncRequestProcessor.java:
```diff
@@ -117,49 +119,41 @@ public void run() {
+                                };
+                            snapInProcess.start();
                         }
-                    } else if (toFlush.isEmpty()) {
```
This is the legacy code which won't be used now, we can get rid of it, but it's better to keep each PR simple with a single purpose, which will make the code review and the future reference easier.

If spotbugs complains things which not caused by the change here, we should leave it as is, and fix it in another PR.

jhuan31 replied:
PR 853

lvfangmin commented on zookeeper-server/src/main/java/org/apache/zookeeper/server/ServerMetrics.java:
```diff
@@ -74,6 +74,16 @@
+    SYNC_PROCESSOR_QUEUE_TIME(new AvgMinMaxPercentileCounter("sync_processor_queue_time_ms")),
+    SYNC_PROCESSOR_QUEUE_SIZE(new AvgMinMaxCounter("sync_processor_queue_size")),
+    SYNC_PROCESSOR_QUEUED(new SimpleCounter("sync_processor_request_queued")),
+    SYNC_PROCESSOR_REAL_QUEUE_TIME(new AvgMinMaxPercentileCounter("sync_processor_real_queue_time_ms")),
```
Let's remove this one, and change SYNC_PROCESSOR_QUEUE_TIME to track the real queue time.

jhuan31 replied:
done

jhuan31 commented:
There are no failing tests. The failure in Jenkins is the findbugs issue "redundant null check" and "unreachable code", which was fixed then reverted according to Fangmin's comment and is now addressed in PR 853

eolivelli commented:
So we should merge #853 before this patch.
It is weird that the maven build is not having problems.
But this is due to the fact that we are using sporbugs

anmolnar commented on zookeeper-server/src/test/java/org/apache/zookeeper/server/quorum/SyncRequestProcessorMetricTest.java:
```diff
@@ -0,0 +1,90 @@
+        syncProcessor.start();
+
+
+        Thread.sleep(500);
```
Sign of a flaky maybe?

jhuan31 replied:
Changed to use a countDown latch. Thanks!

jhuan31 commented:
Added a metric SYNC_PROCESSOR_QUEUE_AND_FLUSH_TIME, which is the end to end time a request stays in the SyncRequestProcessor.  Also changed sleep to CountDownLatch in the unit test.

eolivelli commented on zookeeper-server/src/test/java/org/apache/zookeeper/server/quorum/SyncRequestProcessorMetricTest.java:
```diff
@@ -0,0 +1,102 @@
+        Assert.assertEquals(500L, values.get("sync_processor_request_queued"));
+
+        allRequestsFlushed = new CountDownLatch(500);
+        syncProcessor.start();
```
Very last nit: we are not closing/stopping it. 
So we will leak a few test resources.

eolivelli commented on zookeeper-server/src/test/java/org/apache/zookeeper/server/quorum/SyncRequestProcessorMetricTest.java:
```diff
@@ -0,0 +1,102 @@
+        allRequestsFlushed = new CountDownLatch(500);
+        syncProcessor.start();
+
+        allRequestsFlushed.await(500, TimeUnit.MILLISECONDS);
```
We can use a greater timeout, in order to deal with temporary overloaded test machines.
I suggest 5000

eolivelli commented:
retest ant build

jhuan31 commented:
> Did you see my last comments?

Sorry! Was on vacation. Just made the changes. Thank you @eolivelli

jhuan31 commented:
reset maven build

jhuan31 commented:
retest maven build

jhuan31 commented:
retest maven build
